### PR TITLE
fix cpu utilization

### PR DESCRIPTION
**Pull request by pfrcks**
- use /proc/stat to calculate cpu utilization instead of /proc/uptime

### CODE REVIEW

File: source/code/plugins/VMInsightsDataCollector.rb [hunk 1]
```diff
@@ -54,23 +54,28 @@ def get_available_memory_kb
...
             idle = nil
-            File.open(File.join(@root, "proc", "uptime"), "rb") { |f|
+            File.open(File.join(@root, "proc", "stat"), "rb") { |f|
                 line = f.gets
-                next if line.nil?
-                line.scanf(" %f %f ") { |u, i|
-                    uptime = u
-                    idle = i
+                raise Unavailable, "/proc/stat empty" if line.nil?
+                a = line.split(" ")
```
Comment (jatakiajanvi12) [inline source/code/plugins/VMInsightsDataCollector.rb:72]: "a" doesn't seem an intuitive name for a variable. Can you change to something meaningful?

**Review by jatakiajanvi12 — APPROVED** [review]

**Review by hestolz — APPROVED** [review]

File: source/code/plugins/VMInsightsDataCollector.rb [hunk 2]
```diff
@@ -54,23 +54,28 @@ def get_available_memory_kb
...
-                next if line.nil?
-                line.scanf(" %f %f ") { |u, i|
-                    uptime = u
-                    idle = i
+                raise Unavailable, "/proc/stat empty" if line.nil?
+                time_entries = line.split(" ")
+                # cpu user nice system idle - remaining entries depend on kernel version
+                raise Unavailable, "/proc/stat insufficient entries" if time_entries.length < 5
+                time_entries = time_entries.slice(1, time_entries.length) # skip the first entry in row: "cpu"
+                idle, total_time = time_entries[3].to_i, time_entries.map(&:to_i).sum
```
Comment (mthompson83) [inline source/code/plugins/VMInsightsDataCollector.rb:76]: This implementation does not match [get_cpu_time](https://msazure.visualstudio.com/InfrastructureInsights/_git/DependencyAgent?path=/collector/service/platform/linux/machine-metrics.cc&version=GBmaster&line=104&lineEnd=105&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contentsl)

Comment (mthompson83) [inline source/code/plugins/VMInsightsDataCollector.rb:76]: If all the underlying kernel counters are 64-bit as assumed in [get_cpu_time](https://msazure.visualstudio.com/InfrastructureInsights/_git/DependencyAgent?path=/collector/service/platform/linux/machine-metrics.cc&version=GBmaster&line=104&lineEnd=105&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contentsl), and Jiffies are on the order of 1/100th of a second, then rollover isn't a concern. The C++ implementation implicitly will handle rollover, but the ruby implementation does not.

File: test/code/plugins/VMInsightsDataCollector_subtest.rb [hunk 3]
```diff
@@ -386,34 +400,43 @@ def test_get_filesystems_live
...
+        end
+
+        def test_proc_stat_garbage
+            File.open(@proc_stat, WriteASCII) { |f|
+                f.puts "ga rb age tr ash"
+            }
+            total_time, idle  = @object_under_test.get_cpu_idle
+            assert_equal 0, total_time
+            assert_equal 0, idle
         end
```
Comment (mthompson83) [inline test/code/plugins/VMInsightsDataCollector_subtest.rb:441]: test case for first word on the line is not "cpu" but everything else is ok.

File: test/code/plugins/VMInsightsDataCollector_subtest.rb [hunk 4]
```diff
@@ -386,34 +400,43 @@ def test_get_filesystems_live
...
-            assert ex.message.include?("Uptime not found"), ex.inspect
+            assert ex.message.include?("/proc/stat empty"), ex.inspect
         end
 
-        def test_proc_uptime_garbage
-            File.open(@proc_uptime, WriteASCII) { |f|
-                f.puts "gobble", "d", "gook"
+        def test_proc_stat_insufficient_entries
+            File.open(@proc_stat, WriteASCII) { |f|
+                f.puts "cpu 10 20"
```
Comment (mthompson83) [inline test/code/plugins/VMInsightsDataCollector_subtest.rb:425]: Make this exactly one element less than minimum. Add a test case with exactly the minimum number of elements.

**Review by mthompson83 — CHANGES_REQUESTED** [review]

File: source/code/plugins/VMInsightsDataCollector.rb (hunk 2 above)
Comment (pfrcks) [reply source/code/plugins/VMInsightsDataCollector.rb:76]: not sure I understand. Are you referring to the "* 10000LL" in the C++ code?

**Review by mthompson83 — APPROVED** [review]
